feat: block any RPC request (except from preinstalled snaps) while serving`wallet_requestExecutionPermissions` request

jeffsmale90 opened the pull request:
## **Description**

**WHY:** When serving a `wallet_requestExecutionPermissions` request, multiple confirmations are shown to the user (Permission Picker, [maybe] Upgrade Account, Sign Permission). It's possible for a malicious actor to request a signature immediately after requesting a permission, which results in the signature confirmation being queued after the Permission Picker. To the user, it may appear that this signature is related to the permission being granted, so may unintentionally sign a malicious request.

In order to mitigate the above risk, while serving a `wallet_requestExecutionPermissions` RPC, no other RPC requests will be served, with the exception of requests originating from pre-installed snaps.

Note: this is included only in builds that include `gator-permissions` feature - presently only flask.

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/MetaMask/metamask-extension/pull/37407?quickstart=1)

## **Changelog**

CHANGELOG entry: Block concurrent RPCs while serving `wallet_requestExecutionPermissions`.

## **Manual testing steps**

1. Request `wallet_requestExecutionPermissions` 
2. Any time during the lifecycle of this request, request any other RPC, from the same origin, or another
Expect the subsequent request to be declined with `Cannot process requests while a wallet_requestExecutionPermissions request is in process`

1. Request `wallet_requestExecutionPermissions`
2. Accept or decline the permission - sign the permission if required
3. Make any other RPC, from the same origin, or another
Expect the subsequent request to be served normally 

## **Pre-merge author checklist**

- [x] I've followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Extension Coding Standards](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Blocks concurrent RPCs (except from preinstalled snaps) while processing `wallet_requestExecutionPermissions`, and adds before/after hooks to `forwardRequestToSnap` with robust error handling.
> 
> - **Middleware**:
>   - Add `app/scripts/lib/rpcBlockingMiddleware.ts` to block requests when `state.isBlocked` is true, exempting preinstalled snaps; customizable `errorMessage`. Includes unit tests.
> - **Controller Integration** (`app/scripts/metamask-controller.js`):
>   - Instantiate and push `eip7715BlockingMiddleware` into the EIP-1193 engine.
>   - Toggle blocking via `forwardRequestToSnap` hooks around `wallet_requestExecutionPermissions` (set `isBlocked` true before, false after), with specific error message.
> - **Snap Forwarding** (`app/scripts/lib/forwardRequestToSnap.ts`):
>   - Add `onBeforeRequest` and `onAfterRequest` hooks; ensure `onAfterRequest` runs on success and errors; propagate correct errors. Comprehensive tests added.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit aa51d049c1c17719404a0763ca41467e766c8184. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

jeffsmale90 committed:
Block any RPC request (except from preinstalled snaps) while processing wallet_requestExecutionPermissions request.

jeffsmale90 committed:
Only block requests if gator-permissions feature is enabled (exclude via precompile)

jeffsmale90 committed:
Merge branch 'main' into feat/blockRequestsWhileRequestPermissionsInProgress

jeffsmale90 committed:
Tidy ups from PR
- return provider error when rpc blocking middleware is blocking
- tidy up isSnapPreinstalled mock in rpcBlockingMiddleware tests
- move onBeforeRequest invocation into try block to ensure that onAfterRequest is called if onBeforeRequest throws

jeffsmale90 commented on app/scripts/lib/rpcBlockingMiddleware.ts:
```diff
@@ -0,0 +1,45 @@
+    const { origin } = req;
+
+    if (isBlocked && !isSnapPreinstalled(origin as SnapId)) {
+      end(new Error(errorMessage));
```
Good call!

mj-kiwi commented on app/scripts/lib/forwardRequestToSnap.test.ts:
```diff
@@ -29,6 +29,83 @@ describe('forwardRequestToSnap', () => {
+
+      expect(onBeforeRequest).toHaveBeenCalledTimes(1);
+      expect(handleRequestMock).not.toHaveBeenCalled();
+    });
```
I think we should add a test to cover the following case: when both hooks are provided, `onAfterRequest` is still called even if `onBeforeRequest` throws.

jeffsmale90 replied:
Good call

jeffsmale90 committed:
Merge branch 'main' into feat/blockRequestsWhileRequestPermissionsInProgress

jeffsmale90 committed:
Add additional test to ensure that onAfterRequest is called even if onBeforeRequest throws. Update to JsonRpcMiddlewarev2.

jeffsmale90 committed:
Remove state from RpcBlockingMiddleware (also fix tests to remove end hook)

jeffsmale90 committed:
Merge branch 'main' into feat/blockRequestsWhileRequestPermissionsInProgress

jeffsmale90 committed:
Remove gator-permission code fence. Remove 'end' parameter in test. Correct call to forwardRequestToSnap.

jeffsmale90 committed:
Upgrade rpcblockingmiddleware to jsonrpcmiddleware v2.

jeffsmale90 committed:
Merge branch 'main' into feat/blockRequestsWhileRequestPermissionsInProgress

jiexi commented on app/scripts/lib/rpcBlockingMiddleware.ts:
```diff
@@ -0,0 +1,46 @@
+      }
+
+      if (!isSnapPreinstalled(origin as SnapId)) {
+        throw providerErrors.unauthorized(errorMessage);
```
is this an unauthorized error or something like resourceUnavailable?

jeffsmale90 committed:
Minor improvments:
- throw resourceUnavailable when rpcBlockingMiddleware is blocking requests
- remove unnecessary await from syncronous tests and assertions
- improve internal error message when no origin is specified on the context
- add test to cover case when no origin is specified on the context

jeffsmale90 committed:
Merge branch 'main' into feat/blockRequestsWhileRequestPermissionsInProgress

jeffsmale90 commented on app/scripts/lib/rpcBlockingMiddleware.ts:
```diff
@@ -0,0 +1,46 @@
+      const origin = context.get('origin');
+
+      if (!origin) {
+        throw new InternalError(`No origin specified for method ${req.method}`);
```
Fixed!

jeffsmale90 committed:
Fix lint (import-order) and test (missed context in forwardRequestToSnap call)

mj-kiwi commented on app/scripts/lib/forwardRequestToSnap.ts:
```diff
@@ -41,16 +45,22 @@ export async function forwardRequestToSnap(
+
+    return response;
+  } finally {
+    onAfterRequest?.();
```
I think if `onAfterRequest()` throws in the finally block, it will completely mask any error from `onBeforeRequest()` or `handleRequest()`, making debugging not easy.

jeffsmale90 committed:
Downgrade to middleware v1,
Improve error handling in forwardRequestToSnap to ensure when multiple errors are thrown, all details are returned to the caller

jeffsmale90 committed:
Add comment explaining limitation of isBlocked state variable.

jeffsmale90 committed:
Add rpcBlockingMiddleware to multichain provider. Fix tsdoc comment in rpcBlockingMiddleware.

mj-kiwi reviewed (commented):
Looks good to me. I just have one question I’m not fully sure about.

jeffsmale90 committed:
Import snapid as type only. Remove rpc blocking from multichain api.

jiexi commented on app/scripts/metamask-controller.js:
```diff
@@ -1167,6 +1177,10 @@ export default class MetamaskController extends EventEmitter {
+            onBeforeRequest: () =>
+              (rpcBlockingMiddlewareState.isBlocked = true),
+            onAfterRequest: () =>
+              (rpcBlockingMiddlewareState.isBlocked = false),
```
```suggestion
            onBeforeRequest: () =>
              { rpcBlockingMiddlewareState.isBlocked = true },
            onAfterRequest: () =>
              { rpcBlockingMiddlewareState.isBlocked = false },
```

don't need to return the value, but doesn't really matter
